waiter/src/waiter/scheduler.clj

```diff
@@ -842,8 +842,54 @@
                                  ([service-id] (retrieve-syncer-state @syncer-state-atom service-id)))}))
 
 ;;
-;; Support for tracking killed instances
+;; Support for tracking killed instance candidates
 ;;
+(defn- cleanup-kill-candidate
+  "Cleans up state for a kill candidate.
+   If candidate was not killed but still failed, we trigger the failed instance callback."
+  [kill-candidates-state instance-id]
+  (let [{:keys [track-failed-instance tracked-states]} (get kill-candidates-state instance-id)]
+    (when (and track-failed-instance
+               (contains? tracked-states :failed)
+               (not (contains? tracked-states :killed)))
+      (log/info "tracking kill candidate as failed instance" {:id instance-id})
+      (track-failed-instance))
+    (dissoc kill-candidates-state instance-id)))
+
+(let [kill-candidates-state-atom (atom {})]
+
+  (defn track-kill-candidate!
+    "Tracks which instances are being used as kill candidates for the provided duration."
+    [instance-id reason duration-ms]
+    (log/info "tracking instance as kill candidate" {:duration-sm duration-ms :id instance-id :reason reason})
+    (swap! kill-candidates-state-atom update-in [instance-id :tracked-states] set/union #{reason})
+    (when (= :prepare-to-kill reason)
+      (when (pos? duration-ms)
+        (async/go
+          (async/<! (async/timeout duration-ms))
+          (log/info "end tracking instance as kill candidate" {:id instance-id})
+          (swap! kill-candidates-state-atom cleanup-kill-candidate instance-id)))))
+
+  (defn track-failed-kill-candidate!
+    "Tags a kill candidate as a failed instance for processing during cleanup."
+    [instance-id track-failed-instance-callback]
+    (let [kill-candidate?-atom (atom false)]
+      (swap! kill-candidates-state-atom
+             (fn update-failed-kill-candidate-state-1 [kill-candidates-state]
+               (cond-> kill-candidates-state
+                 (contains? kill-candidates-state instance-id)
+                 (update instance-id (fn update-failed-kill-candidate-state-2 [state]
+                                       (reset! kill-candidate?-atom true)
```

Comment:
From the [docs for `swap!`](https://clojuredocs.org/clojure.core/swap!):

> Note that `f` may be called multiple times, and thus should be free of side effects.  Returns the value that was swapped in.

Rather than using the atom, should we just look at the value returned from `swap!` to see if it has `instance-id :tracked-states :failed`?

Reply:
Removed this function.